src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
+                }
+                if (_notifyFilters.HasFlag(NotifyFilters.FileName))
+                {
+                    _filterFlags |= Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsFile | 
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemIsSymlink |
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemCreated |
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemModified |
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRemoved |
+                                    Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagItemRenamed;
+                }
+
+                // The directory has not changed then just restart receiving events. If the directory has changed 
+                // then we need to cleanup the old stream and kick off a new one
+                if (_lastWatchedDirectory.Equals(_directory, StringComparison.CurrentCultureIgnoreCase))
+                {
+                    RestartStream();
+                }
+                else
+                {
+                    CleanupStreamAndWatcher();
+                    CreateStreamAndStartWatcher();
+                    _lastWatchedDirectory = _directory;
+                }
+            }
+        }
+
+        private void CreateStreamAndStartWatcher()
+        {
+            Debug.Assert(_eventStream == IntPtr.Zero);
+            Debug.Assert(_watcherRunLoop == IntPtr.Zero);
+
+            // Make sure we only do this if there is a valid directory
+            if (String.IsNullOrEmpty(_directory) == false)
+            {
+                // Get the path to watch and put it into a CFArray.
+                CFStringRef path = Interop.CoreFoundation.CFStringCreateWithCString(IntPtr.Zero, _directory, Interop.CoreFoundation.CFStringBuiltInEncodings.kCFStringEncodingUTF8);
+                CFArrayRef arrPaths = Interop.CoreFoundation.CFArrayCreate(IntPtr.Zero, new CFStringRef[1] { path }, 1, IntPtr.Zero);
+
+                // Create the event stream for the path and tell the stream to watch for file system events.
+                _eventStream = Interop.CoreServices.FSEventStreamCreate(
+                    IntPtr.Zero,
+                    FileSystemEventCallback,
```

Comment:
You're passing a method group here.  That'll cause the compiler to allocate a delegate to wrap that method, and the runtime will effectively pin that delegate for the duration of this interop call, but as soon as Create returns, that delegate can be collected.  At some point in the future, GC can run, cleanup that delegate, and then you'll have native code trying to access a delegate that no longer exists.  To fix this, create a field to store the delegate, create a delegate for this method and store it into that field, and then pass that same object into this call.